### PR TITLE
Fix Segfault in EVP_PKEY_CTX_dup when the ctx has an undefined operation.

### LINKED ISSUE #12438: EVP_PKEY_CTX_dup Segmentation fault
```
#0  EVP_PKEY_CTX_dup (pctx=0x5555555b5500) at ../crypto/evp/pmeth_lib.c:516
#1  0x0000555555555299 in main ()
    at main.cc:32
```

```
#include <openssl/ec.h>
#include <openssl/err.h>
#include <openssl/evp.h>

int main() {
    EVP_PKEY_CTX *pctx{nullptr};
    if (!(pctx = ::EVP_PKEY_CTX_new_id(EVP_PKEY_EC, nullptr)))
        throw;

    if (::EVP_PKEY_paramgen_init(pctx) <= 0)
        throw;

    if (::EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pctx, NID_X9_62_prime256v1) <= 0)
        throw;

    EVP_PKEY *param{nullptr};
    if (::EVP_PKEY_paramgen(pctx, &param) <= 0)
        throw;

    if (!param)
        throw;

    ::EVP_PKEY_CTX_free(pctx);

    EVP_PKEY_CTX *ctx{nullptr};

    if (!(ctx = ::EVP_PKEY_CTX_new(param, nullptr))) {
        throw;
    }
    
    EVP_PKEY_CTX *kctx{nullptr};
    if (!(kctx = ::EVP_PKEY_CTX_dup(ctx)))
        throw;
    return 0;
}
```
openssl version master:a01cae99ac384cb6a74b46ccdc90736fe0754958

### PR DESCRIPTION
**Pull request by slontis**
Fixes #12438

Note: This worked in 1.1.1 so just returning an error is not valid.

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
@levitte I am not sure if this is the correct way to dup the keymt pointer.. I have tried to do something similiar to what other defines operations do.

File: crypto/evp/pmeth_lib.c [hunk 1]
```diff
@@ -475,105 +475,117 @@ EVP_PKEY_CTX *EVP_PKEY_CTX_dup(const EVP_PKEY_CTX *pctx)
...
-                OPENSSL_free(rctx);
-                return NULL;
-            }
+            if (!EVP_KEYEXCH_up_ref(rctx->op.kex.exchange))
+                goto end;
         }
         if (pctx->op.kex.exchprovctx != NULL) {
             if (!ossl_assert(pctx->op.kex.exchange != NULL))
-                return NULL;
+                goto end;
```
Comment (slontis) [inline crypto/evp/pmeth_lib.c:488]: This one and others were a potential memory leak..

File: crypto/evp/pmeth_lib.c [hunk 2]
```diff
@@ -475,105 +475,117 @@ EVP_PKEY_CTX *EVP_PKEY_CTX_dup(const EVP_PKEY_CTX *pctx)
...
             if (rctx->op.encap.kemprovctx == NULL) {
                 EVP_KEM_free(rctx->op.encap.kem);
-                OPENSSL_free(rctx);
-                return NULL;
+                goto end;
             }
             return rctx;
         }
+    } else if (EVP_PKEY_CTX_IS_GEN_OP(pctx)) {
+        /* Not supported - This would need a gen_dupctx() to work */
```
Comment (slontis) [inline crypto/evp/pmeth_lib.c:549]: Do we need this??

Comment (t8m) [reply crypto/evp/pmeth_lib.c:549]: Perhaps not, but should it be explicitly documented as not supported?

Comment (slontis) [reply crypto/evp/pmeth_lib.c:549]: Added a line to the docs.

**Comment by slontis** [comment]
rebased

**Comment by slontis** [comment]
ping

**Review by kaduk — APPROVED** [review]
I'm not fully up to speed with all of this stuff (e.g., evp_pkey_export_to_provider()) but the memory-leak fixes are right, and I convinced myself that it's not harmful and probably addresses the indicated issue.  (Also, it does fix the test case it added.)
Approved, with nit fix to be made while merging.

File: crypto/evp/pmeth_lib.c [hunk 3]
```diff
@@ -488,105 +488,117 @@ EVP_PKEY_CTX *EVP_PKEY_CTX_dup(const EVP_PKEY_CTX *pctx)
...
 
-    if (pctx->peerkey)
+    if (pctx->peerkey != NULL)
         EVP_PKEY_up_ref(pctx->peerkey);
     rctx->peerkey = pctx->peerkey;
 
+    if (pctx->pmeth == NULL) {
+        if (rctx->operation == EVP_PKEY_OP_UNDEFINED) {
+            EVP_KEYMGMT *tmp_keymgmt = NULL;
+            void *provkey = NULL;
```
[the remark this answers is not in the document]
Comment (slontis) [reply crypto/evp/pmeth_lib.c]: same applies to tmp_keymgmt

**Comment by slontis** [comment]
Updated to address @kaduk comments.

**Review by kaduk — APPROVED** [review]
Reconfirm +1 -- thanks for the fixups

**Comment by slontis** [comment]
Thanks. Merged to master.

**Pull request closed without merging**